vmaas_sync/metrics_cyndi.go

```diff
@@ -0,0 +1,165 @@
+	"fmt"
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"reflect"
+	"time"
+)
+
+var (
+	cyndiSystemsCnt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "cyndi_systems",
+	}, []string{"type"})
+
+	cyndiTagsCnt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "cyndi_tags_count",
+	}, []string{"type"})
+)
+
+type InventoryHosts []struct {
+	DisplayName   string
+	Tags          string
+	SystemProfile string
+}
+
+type Tags []struct {
+	Key       string `json:"key"`
+	Value     string `json:"value"`
+	Namespace string `json:"namespace"`
+}
+
+type SystemProfile struct {
+	SapSystem bool `json:"sap_system"`
+}
+
+//Firstly, this method unmarshal jsonb fields from DB, and loads them into relevant struct
+//This method also exposes metrics regarding max/min tags per system and number of systestems
+//with/without tags and counts systems with sap_system=true
+func updateCyndiData() {
+	count, err := getCyndiData()
+	if err != nil {
+		utils.Log("err", err.Error()).Error("unable to update cyndi metrics")
+		return
+	}
+
+	for key, value := range count {
+		cyndiTagsCnt.WithLabelValues(key).Set(float64(value))
+	}
+	// cyndiTagsCnt.WithLabelValues("uniqueTagsCount").Set(float64(count["uniqueTagsCount"]))
+	// cyndiTagsCnt.WithLabelValues("systemsCount").Set(float64(count["allSystemCount"]))
+	// cyndiTagsCnt.WithLabelValues("systemsWithSap").Set(float64(count["systemsSapSystem"]))
+	// cyndiTagsCnt.WithLabelValues("systemsWithTags").Set(float64(count["systemsWithTag"]))
+}
+
+//Loads display_name, tags and system_profile from inventory.hosts
+//and loads them in InventoryHosts struct
+func getCyndiData() (map[string]int, error) {
+	count := map[string]int{}
+	uniqueTags := []Tags{}
+	var stats InventoryHosts
+	err := database.Db.Table("inventory.hosts").Scan(&stats).Error
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get inventory hosts from db")
+	}
+
+	var systemsWithTag int = 0
+	var systemsSapSystem int = 0
+	count["allSystemCount"] = len(stats)
+	for _, item := range stats {
+		var tags Tags
+		err := json.Unmarshal([]byte(item.Tags), &tags)
+		if err != nil {
+			utils.Log("err", err.Error()).Error("Not able to unmarshall")
+		}
+		uniqueTags = append(uniqueTags, tags)
+		var systemProfile SystemProfile
+		err = json.Unmarshal([]byte(item.SystemProfile), &systemProfile)
+		if err != nil {
+			utils.Log("err", err.Error()).Error("Not able to unmarshall")
+		}
+		if tags != nil {
+			systemsWithTag++
+			maxTag := len(tags)
+			count[item.DisplayName] = maxTag
+		}
+		if systemProfile.SapSystem {
+			systemsSapSystem++
+		}
+	}
+	count["uniqueTagsCount"] = len(unique(uniqueTags))
+	count["systemsSapSystem"] = systemsSapSystem
+	count["systemsWithTag"] = systemsWithTag
+
+	return count, nil
+}
+
+func unique(inputTags []Tags) []Tags {
```

Comment:
The convention is to have a `map[Tag]bool` and insert all values into this map, then create a new slice from the map.